docs/spec/annotations.rst

```diff
@@ -53,43 +53,154 @@ decorators ``@property``, ``@staticmethod`` and ``@classmethod``.
 
 .. _valid-types:
 
-Valid type expression forms
----------------------------
-
-Type hints may be built-in classes (including those defined in
-standard library or third-party extension modules), abstract base
-classes, types available in the ``types`` module, and user-defined
-classes (including those defined in the standard library or
-third-party modules).
-
-While annotations are normally the best format for type hints,
-there are times when it is more appropriate to represent them
-by a special comment, or in a separately distributed stub
-file.  (See below for examples.)
+Type and annotation expressions
+-------------------------------
+
+The terms *type expression* and *annotation expression* denote specific
+subsets of Python expressions that are used in the type system.
+
+.. _`type-expression`:
+
+A *type expression* is any expression that validly expresses a type. Type
+expressions are always acceptable in annotations and also in various other
+places. Specifically, type expressions are used in the following locations:
+
+* In a type annotation (always as part of an annotation expression)
+* The first argument to :ref:`cast() <cast>`
+* The second argument to :ref:`assert_type() <assert-type>`
+* The bounds and constraints of a ``TypeVar`` (whether created through the
+  old syntax or the native syntax in Python 3.12)
+* The definition of a type alias (whether created through the ``type`` statement,
+  the old assignment syntax, or the ``TypeAliasType`` constructor)
+* The type arguments of a generic class (which may appear in a base class
+  or in a constructor call)
+* The definitions of fields in the class-based forms for creating
```

Comment:
I think you mean the functional forms, right? The class-based forms use proper variable annotations.

Reply:
Yes, thanks